spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/logging/StackdriverJsonLayout.java

```diff
@@ -163,6 +165,17 @@ public void setIncludeExceptionInMessage(boolean includeExceptionInMessage) {
 		this.includeExceptionInMessage = includeExceptionInMessage;
 	}
 
+	@Override
+	public void start() {
+		super.start();
+
+		// If no Project ID set, then attempt to resolve it with the default project ID provider
+		if (StringUtils.isEmpty(this.projectId) || this.projectId.endsWith("_IS_UNDEFINED")) {
```

Comment:
Why do we do this?

If there is no configured project ID, should't we ask the user to provide one before jumping to auto-configuration?

Reply:
this is pretty similar to autoconfigurations. setting project ID per deployment/project is quite a hassle. we don't require it for the rest of spring cloud gcp.  i feel it should at least attempt to resolve it via the standard way.  If they need to use spring.cloud.gcp.project-id, they'll need to use `<springProperty/>`

Reply:
Auto-configuration already supports credential discovery, like you're doing it here.

It's the last resort, and it's done through `DefaultCredentialsProvider`.

Which is why I think we should avoid this code duplication.

Reply:
This is because we have no access to the Spring context from the Layout :(